### PR TITLE
Sync phone changes with WooPay phone field

### LINKED ISSUE #9235: WooPay opt-in changes to wrong country when clicking in then out after autocomplete
### Describe the bug
The WooPay opt-in changes the country code to 1 after clicking in then out of the phone field after it was auto completed by copying the billing phone. **The phone is correctly saved** but the wrong flag shows up during checkout.

https://github.com/user-attachments/assets/146e9656-105d-45dd-a579-cd5160694cae

### To Reproduce
- As a new unlogged customer add a product to your cart
- Go to the blocks checkout page
- Fill in the billing details
- Use a billing phone from another country like `+5541991788166`
- Click the WooPay opt-in checkbox (Securely save my information for 1-click checkout)
- It should copy the billing phone
- Click the phone field then click out
- The country flag will change to US but it should stay as Brazil
- The WooPay phone number was saved correctly in the database but the flag shows the wrong country during checkout

### LINKED ISSUE #9431: Save my info component phone number is still not pre-filled
### Describe the bug
On blur of the phone number field, the WooPay save my info phone number is not pre-filled. It's only ever pre-filled if the shopper unchecks, and re-checks the checkbox.

https://github.com/user-attachments/assets/16fb0081-23db-43c2-bbc4-61060b3ae657

### Acceptance criteria
1. On `blur` or `focusout` of the address phone number, if the phone number is populated and it's valid, pre-fill the WooPay save my info phone number
2. On checking of the save my info checkbox, if a valid address phone number is available, pre-fill the WooPay component.
3. In countries where pre-checking is not enabled, on checking of the save my info checkbox, if a valid address phone number is available, pre-fill the WooPay component.

### LINKED ISSUE #9463: WooPay phone fields does not get updated when the billing phone is changed
### Describe the bug

Currently when the phone field is changed and WooPay `Save my info` is checked, the WooPay phone number field does not get update.

### To Reproduce

1. Enable WooPay
2. Add products to the cart
3. Check `Securely save my information for 1-click checkout`.
4. Change the billing address phone number.
5. The WooPay phone field does not get updated.

### Actual behavior

WooPay phone field does not update when the billing phone is updated.

### Expected behavior

WooPay phone field get updated when the billing phone is updated.

### PR DESCRIPTION
**Pull request by alefesouza**
Fixes #9431
Fixes #9235

#### Changes proposed in this Pull Request

When removing the focus of the billing address phone field, automatically update the WooPay phone number. 

#### Testing instructions

* Enable WooPay.
* Add products to the cart.
* Access blocks checkout page.
* Check `Securely save my information for 1-click checkout`.
* Change the billing address phone number and remove focus.
* The WooPay phone field should get updated.
* Test on classic checkout.
* Test again with WooPay pre-checking enabled (WCPay DevTools > `Force the pre_check_save_my_info flag (aka Default Opt-in) in the account cache to be true`).

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9464` or branch name `fix/sync-woopay-phone-field` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/sync-woopay-phone-field) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 13718c82861430e94c10120f6a04d72048c15d1b
- Build time: 2024-09-27 04:30:43 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Review by leonardola — CHANGES_REQUESTED** [review]
When I type a Brazilian phone number the country code changes back to US when the input loses focus

https://github.com/user-attachments/assets/45308ed2-b724-42b3-9012-a99bf4ffc724

When I autofill the opt-in phone number the page jumps to the billing address fields
When I autofill the billing address after I typed the opt-in it gets removed

https://github.com/user-attachments/assets/0b47a567-71f7-417f-bfc5-3c1735f503e5

When the number gets copied from the billing address it does not change the country flag

https://github.com/user-attachments/assets/1c679412-e625-479a-9555-7603ac7ff678

**Comment by alefesouza** [comment]
@leonardola fixed on https://github.com/Automattic/woocommerce-payments/pull/9464/commits/9b3b9910d04c8abc90d071b7af2bbd42c7f75b45.

File: client/components/woopay/save-user/checkout-page-save-user.js [hunk 1]
```diff
@@ -222,6 +237,20 @@ const CheckoutPageSaveUser = ( { isBlocksCheckout } ) => {
...
+		if ( isCustomerDataUpdating ) {
+			setTimeout( () => {
+				setPhoneNumber( getPhoneFieldValue() );
+			}, 0 );
+		}
+	}, [
+		billingAddressPhone,
+		setPhoneNumber,
+		getPhoneFieldValue,
+		isCustomerDataUpdating,
```
Comment (alefesouza) [inline client/components/woopay/save-user/checkout-page-save-user.js:251]: Used this approach due to inconsistencies when using the `blur` event and autofill.

Comment (alefesouza) [inline client/components/woopay/save-user/checkout-page-save-user.js:251]: Used this approach due to inconsistencies when using the blur event and autofill.

File: client/components/woopay/save-user/checkout-page-save-user.js [hunk 2]
```diff
@@ -222,6 +237,20 @@ const CheckoutPageSaveUser = ( { isBlocksCheckout } ) => {
...
+	] );
```
Comment (alefesouza) [inline client/components/woopay/save-user/checkout-page-save-user.js]: Used this approach due to inconsistencies when using the blur event and autofill.

**Review by leonardola — CHANGES_REQUESTED** [review]
Is it expected to update the billing/shipping phone number when the WooPay phone is first typed/autofilled but then not update it if it changes?

Wouldn't it cause problems because the customer might have mistyped it the first time and tried to fix it?

Also after autofilling the WooPay number the page jumps to the billing/shipping section without any smooth transition

https://github.com/user-attachments/assets/26d40edf-d644-4d8b-b16b-d2e95345e3af

Is it expected that the WooPay phone number will always be updated when the billing/shipping phone is changed?

Won't this cause problems when people want one phone number to login to WooPay and another one for shipping?

https://github.com/user-attachments/assets/088ebe8d-70af-4a6b-b88f-54419207d486

Should we only copy the shipping phone number or should we copy the billing phone number instead when available?

I think it would make more sense to copy the billing phone number instead of the shipping phone number when available

https://github.com/user-attachments/assets/66f665a8-c67b-4961-952c-3737848cb127

https://github.com/user-attachments/assets/392aafa6-7db4-4cc6-beed-57e92f755210

**Comment by alefesouza** [comment]
@leonardola fixed the field getting update after being changed on https://github.com/Automattic/woocommerce-payments/pull/9464/commits/4d8bee60c9aa5759cf74a895b8f46f323310a347 and the billing address preference on https://github.com/Automattic/woocommerce-payments/pull/9464/commits/c7766f764501322d0859cd47d7f921c1a54862d1 , the autocomplete field is not related to this PR and appears on current `develop`, could you open an issue for that?

**Review by leonardola — CHANGES_REQUESTED** [review]
Somehow it is copying the whole Shipping address to the billing address fields after I use autofill. 

Notice that I use an American address in the shipping fields and a Brazilian address on billing address. But then the billing gets updated with a copy of the shipping address and the WooPay phone gets updated with the Brazilian phone number.

Steps used:
- Add a product to your cart
- Go to the checkout page
- Fill in a Shipping address
- Enable WooPay opt-in
- Change the Billing address
- The billing address will get updated with the shipping address
- The WooPay number will be the one you used in the billing address before it changed

https://github.com/user-attachments/assets/9a515f31-73e2-440f-8da9-b53036ba9486

When I change the billing number it gets reverted to the previous number and the WooPay phone number gets updated to the number I just typed

Steps used:
- Add a product to your cart
- Go to the checkout page
- Fill in a Shipping address
- Fill in the Billing address
- Enable WooPay opt-in
- Change the Billing address phone
- Notice the Billing address will be reverted
- Notice that the WooPay number will be the one you just typed in the billing address before it changed back

https://github.com/user-attachments/assets/986c5f6b-bccc-4cc9-b141-4df506aebac1

**Comment by alefesouza** [comment]
I created a new endpoint for the WooPay phone number update on https://github.com/Automattic/woocommerce-payments/pull/9464/commits/a079b3c2ad609b9c61da3703145ea50bce25940d because using the `extensionCartUpdate` right after the address update would create a racing condition, there will be a Store API request to update the customer address and another to update the WooPay phone number, when running the WooPay request from autofill the address is still empty because the first request is not finished yet, the same happens when clicking billing same as shipping checkbox, because of all of that the cart frontend would update to the empty `extensionCartUpdate` request instead of the autofill update:

| Request    | Response |
| -------- | ------- |
|  Screenshot 2024-09-26 at 9 03 01 PM |  Screenshot 2024-09-26 at 9 03 25 PM |

I think the checkout block frontend applies the first response, causing the bugs we had seen before, by not using the Store API to save the WooPay phone number we prevent this racing condition and fix this problem.

File: client/components/woopay/save-user/checkout-page-save-user.js [hunk 3]
```diff
@@ -114,22 +117,28 @@ const CheckoutPageSaveUser = ( { isBlocksCheckout } ) => {
...
 						woopay_user_phone_field: {
 							full: phoneNumber,
 						},
 				  };
 
-			extensionCartUpdate( {
-				namespace: 'woopay',
-				data: data,
-			} )?.then( () => {
+			request(
```
Comment (leonardola) [inline client/components/woopay/save-user/checkout-page-save-user.js:132]: Why did you switch the `extensionCarUpdated` with an ajax call?

File: includes/woopay/class-woopay-utilities.php [hunk 4]
```diff
@@ -123,10 +122,10 @@ public function should_save_platform_customer() {
 		$session_data = [];
 
 		if ( isset( WC()->session ) && method_exists( WC()->session, 'has_session' ) && WC()->session->has_session() ) {
-			$session_data = WC()->session->get( WooPay_Extension::WOOPAY_SESSION_KEY );
+			$session_data = WC()->session->get( WooPay_Session::WOOPAY_SESSION_KEY );
 		}
 
-		return ( isset( $_POST['save_user_in_woopay'] ) && filter_var( wp_unslash( $_POST['save_user_in_woopay'] ), FILTER_VALIDATE_BOOLEAN ) ) || ( isset( $session_data['save_user_in_woopay'] ) && filter_var( $session_data['save_user_in_woopay'], FILTER_VALIDATE_BOOLEAN ) ); // phpcs:ignore WordPress.Security.NonceVerification
+		return isset( $session_data['save_user_in_woopay'] ) && filter_var( $session_data['save_user_in_woopay'], FILTER_VALIDATE_BOOLEAN );
```
Comment (leonardola) [inline includes/woopay/class-woopay-utilities.php:128]: Why we're ignoring $_POST now?

**Review by leonardola — APPROVED** [review]
Tests ok. I left a couple comments but it's good to merge already

File: client/components/woopay/save-user/checkout-page-save-user.js (hunk 3 above)
Comment (alefesouza) [reply client/components/woopay/save-user/checkout-page-save-user.js:132]: It's explained in my [latest comment](https://github.com/Automattic/woocommerce-payments/pull/9464#issuecomment-2378361958).

File: includes/woopay/class-woopay-utilities.php (hunk 4 above)
Comment (alefesouza) [reply includes/woopay/class-woopay-utilities.php:128]: Using the ajax call we are always saving in session, so there's no need for the `$_POST`.

**Pull request merged**